### ISSUE TITLE
Sparse GPR (Titsias, 2009)

### ISSUE THREAD

**Issue by vr308** [cornellius-gp/gpytorch#1585]
I am trying to run a toy example for the Titsias Sparse GPR where the objective is the collapsed bound and inducing inputs are learnt along with kernel hypers. What is a bit non-standard is that gpytorch demo subclasses the ExactGP and just sees this whole thing through the InducingPointKernel().

```
import gpytorch 
import torch
import numpy as np
from prettytable import PrettyTable
from gpytorch.means import ConstantMean
from gpytorch.kernels import ScaleKernel, RBFKernel, InducingPointKernel
from gpytorch.distributions import MultivariateNormal



class SparseGPR(gpytorch.models.ExactGP):
    
    def __init__(self, train_x, train_y, likelihood):
        
        """The sparse GP class for regression with the collapsed bound.
        """
        super(SparseGPR, self).__init__(train_x, train_y, likelihood)
        self.mean_module = ConstantMean()
        self.base_covar_module = ScaleKernel(RBFKernel())
        self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=train_x[:500], likelihood=likelihood)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)

if __name__ == "__main__":

   def func(x):
      return np.sin(x * 3 * 3.14) + 0.3 * np.cos(x * 9 * 3.14) + 0.5 * np.sin(x * 7 * 3.14)
  
   N = 10000  # Number of training observations
  
   X = torch.randn(N) * 2 - 1  # X values
   Y = func(X) + 0.2 * torch.randn(N)  # Noisy Y values

  likelihood = gpytorch.likelihoods.GaussianLikelihood()
  model = SparseGPR(X, Y, likelihood)
  
  # Find optimal model hyperparameters
  model.train()
  likelihood.train()
  
  # Use the adam optimizer
  optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
  
  # "Loss" for GPs - the marginal log likelihood
  mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
  
  for i in range(1000):
    # Zero backprop gradients
    optimizer.zero_grad()
    # Get output from model
    output = model(X)
    # Calc loss and backprop derivatives
    loss = -mll(output, Y)
    loss.backward()
    #print('Iter %d/%d - Loss: %.3f' % (i + 1, 1000, loss.item()))
    print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f   noise: %.3f' % (
          i + 1, 1000, loss.item(),
         model.base_covar_module.base_kernel.lengthscale.item(),
          model.likelihood.noise.item()))
    optimizer.step()
```
  
Going through with the usual motions and somehow the loss doesn't budge and neither do the hypers.

Iter 1/1000 - Loss: 1.262   lengthscale: 0.693   noise: 0.693
Iter 2/1000 - Loss: 1.262   lengthscale: 0.693   noise: 0.693
Iter 3/1000 - Loss: 1.262   lengthscale: 0.693   noise: 0.693
Iter 4/1000 - Loss: 1.262   lengthscale: 0.693   noise: 0.693
Iter 5/1000 - Loss: 1.262   lengthscale: 0.693   noise: 0.694
Iter 6/1000 - Loss: 1.262   lengthscale: 0.693   noise: 0.694

Clearly something blatantly off with my example?

Next, because the variational distribution q(u) is implicit in this formulation - I guess its not computed anywhere and if I must access it, I must code up the closed form optimal Guassian dist representing q(u)?

**Comment by gpleiss** [comment]
> Clearly something blatantly off with my example?

A few thoughts:
- Use `torch.optim.Adam` instead of `torch.optim.SGD`.
- Based on `func`, I think the characteristic lengthscale of your problem is quite small, and so it's easy to get the optimization caught in a spurious local minimum (also this might not be a good function for SGPR). When I changed `func` to `np.sin(x * 3 * 3.14) + 0.3 * np.cos(x * 4 * 3.14) + 0.5 * np.sin(x * 2 * 3.14)`, optimization was better.

> If I must access it, I must code up the closed form optimal Guassian dist representing q(u)?

A simple solution would be to call `model(model.covar_module.inducing_points)`.

**Comment by vr308** [comment]
@gpleiss thank you, yes those suggestions make a difference. 

Further, I noticed it is better to initalise inducing points randomly rather than at a subset of the training points. In the latter, they dont' seem to budge post training. Also mentioned in this issue: https://github.com/cornellius-gp/gpytorch/issues/1380

**Comment by vr308** [comment]
SGPR predictive posterior variance:

For some reason I am seeing negative variances in the diagonal of the final predictive posterior (this is my script that can just be run to reproduce the issue).

I diagnosed this when I noticed that the posterior predictive variance was vanishingly small in regions where there is a lot of data. So, since this is the SGPR inducing point framework, the predictive variance is:

pred_covar = K_ss - K_sm*Kmm^{-1}K_ms + K_sm*K_mm^{-1}\Sigma K_mm^{-1}K_ms

where K_ss is the covariance between test points, K_sm is the covariance between test and inducing, \Sigma is the covariance of the optimised q*(u) \sim N(\mu, \Sigma).

![Figure_1](https://user-images.githubusercontent.com/17442913/116449171-cc663600-a851-11eb-84ba-0cdbbbecef5e.png)

```
import gpytorch 
import torch
import numpy as np
from prettytable import PrettyTable
from gpytorch.means import ConstantMean, ZeroMean
from gpytorch.kernels import ScaleKernel, RBFKernel, InducingPointKernel
from gpytorch.distributions import MultivariateNormal
import matplotlib.pyplot as plt

def func(x):
    return np.sin(x * 3) + 0.3 * np.cos(x * 4 * 3.14) 


class SparseGPR(gpytorch.models.ExactGP):
    
    def __init__(self, train_x, train_y, likelihood, Z_init):
        
        """The sparse GP class for regression with the collapsed bound.
           q*(u) is implicit. 
        """
        super(SparseGPR, self).__init__(train_x, train_y, likelihood)
        self.train_x = train_x
        self.train_y = train_y
        self.inducin_points = Z_init
        self.num_inducing = len(Z_init)
        self.mean_module = ZeroMean()
        self.base_covar_module = ScaleKernel(RBFKernel())
        self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=Z_init, likelihood=likelihood)

    def forward(self, x): # returns ?
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)
    
    def elbo(self, output, y):
        
        Knn = model.base_covar_module(self.train_x).evaluate()
        Knm = model.base_covar_module(self.train_x, Z_init).evaluate()
        lhs = torch.matmul(Knm, self.covar_module._inducing_mat.inverse())
        Qnn = torch.matmul(lhs, Knm.evaluate().T)
        p_y = model.likelihood(output)
        expected_log_lik = p_y.log_prob(y)
        shape = Knn.shape[:-1]
        noise_diag = self.likelihood._shaped_noise_covar(shape).diag()
        diag = Knn.diag() - Qnn.diag()
        trace_term = 0.5*(diag/noise_diag).sum() 
    
        return expected_log_lik, trace_term
            
    def train_model(self, likelihood, combine_terms=True):
        
        self.train()
        likelihood.train()
        mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
        
        losses = []
        for i in range(1000):
          optimizer.zero_grad()
          output = self(self.train_x)
          if combine_terms:
              loss = -mll(output, self.train_y)
          else:
              loss = -self.elbo(output, self.train_y)
          losses.append(loss)
          loss.backward()
          if i%100 == 0:
                    print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f   noise: %.3f' % (
                    i + 1, 1000, loss.item(),
                    self.base_covar_module.base_kernel.lengthscale.item(),
                    likelihood.noise.item()))
          optimizer.step()
        return losses
    
    def optimal_q_u(self):
       return self(self.covar_module.inducing_points)
    
    def posterior_predictive(self, test_x):
        
        ''' Returns the posterior predictive multivariate normal '''
        
        self.eval()
        self.likelihood.eval()

        # Test points are regularly spaced along [0,1]
        # Make predictions by feeding model through likelihood
        with torch.no_grad(), gpytorch.settings.fast_pred_var():
            y_star = likelihood(self(test_x))
        return y_star
    
    def visualise_posterior(self, test_x, y_star):
    
        ''' Visualising posterior predictive '''
        
        f, ax = plt.subplots(1, 1, figsize=(8, 8))
        lower, upper = y_star.confidence_region()
        ax.plot(self.train_x.numpy(), self.train_y.numpy(), 'kx')
        ax.plot(test_x.numpy(), y_star.mean.numpy(), 'b-')
        ax.plot(self.covar_module.inducing_points.detach(), [-2.5]*self.num_inducing, 'rx')
        ax.fill_between(test_x.numpy(), lower.numpy(), upper.numpy(), alpha=0.5)
        ax.set_ylim([-3, 3])
        ax.legend(['Train', 'Mean', 'Inducing inputs', r'$\pm$2\sigma'])
        plt.show()

if __name__ == '__main__':
    
    N = 1000  # Number of training observations

    X = torch.randn(N) * 2 - 1  # X values
    Y = func(X) + 0.2 * torch.randn(N)  # Noisy Y values

    # Initial inducing points
    Z_init = torch.randn(12)
    
    # Initialise model and likelihood
    likelihood = gpytorch.likelihoods.GaussianLikelihood()
    model = SparseGPR(X, Y, likelihood, Z_init)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.05)
        
    # Train
    losses = model.train_model(likelihood, combine_terms=True)

    # Test 
    test_x = torch.linspace(-8, 8, 1000)
    test_y = func(test_x)
    
    y_star = model.posterior_predictive(test_x)
    
    # Visualise 
    
    model.visualise_posterior(test_x, y_star)
    
    # Compute metrics
    rmse = model.rmse(y_star, test_y)
    nll = model.neg_test_log_likelihood(y_star, test_y)
    
    # Plot predictive variances 
    plt.figure()
    plt.plot(y_star.covariance_matrix.diag().detach())
```

**Comment by vr308** [comment]
Closing this issue.
